src/workerd/io/tracer.c++

```diff
@@ -245,70 +235,92 @@ void WorkerTracer::addSpan(CompleteSpan&& span) {
     kj::String message,
     kj::Maybe<kj::String> stack) {
-  if (trace->exceededExceptionLimit) {
-    return;
-  }
   // TODO(someday): For now, we're using logLevel == none as a hint to avoid doing anything
   //   expensive while tracing.  We may eventually want separate configuration for exceptions vs.
   //   logs.
   if (pipelineLogLevel == PipelineLogLevel::NONE) {
     return;
   }
-  size_t newSize = trace->bytesUsed + sizeof(tracing::Exception) + name.size() + message.size();
+
+  size_t messageSize = sizeof(tracing::Exception) + name.size() + message.size();
   KJ_IF_SOME(s, stack) {
-    newSize += s.size();
+    messageSize += s.size();
+  }
+  KJ_IF_SOME(writer, maybeTailStreamWriter) {
+    auto maybeTruncatedName = name.first(kj::min(name.size(), MAX_TRACE_BYTES));
+    auto maybeTruncatedMessage =
+        message.first(kj::min(message.size(), MAX_TRACE_BYTES - maybeTruncatedName.size()));
+    kj::Maybe<kj::String> maybeTruncatedStack;
+    KJ_IF_SOME(s, stack) {
+      maybeTruncatedStack = kj::heapString(s.first(kj::min(
+          s.size(), MAX_TRACE_BYTES - maybeTruncatedName.size() - maybeTruncatedMessage.size())));
+    }
+    writer->report(context,
+        {tracing::Exception(timestamp, kj::str(maybeTruncatedName), kj::str(maybeTruncatedMessage),
+            kj::mv(maybeTruncatedStack))});
+  }
+
+  if (trace->exceededExceptionLimit) {
+    return;
   }
-  if (newSize > MAX_TRACE_BYTES) {
+
+  if (trace->bytesUsed + messageSize > MAX_TRACE_BYTES) {
     trace->exceededExceptionLimit = true;
     trace->truncated = true;
     trace->exceptions.add(timestamp, kj::str("Error"),
         kj::str("Trace resource limit exceeded; subsequent exceptions not recorded."), kj::none);
-    return;
-  }
-  trace->bytesUsed = newSize;
-  KJ_IF_SOME(writer, maybeTailStreamWriter) {
-    writer->report(context,
-        {tracing::Exception(timestamp, kj::str(name), kj::str(message), mapCopyString(stack))});
+  } else {
+    trace->bytesUsed += messageSize;
+    trace->exceptions.add(timestamp, kj::mv(name), kj::mv(message), kj::mv(stack));
   }
-  trace->exceptions.add(timestamp, kj::mv(name), kj::mv(message), kj::mv(stack));
 }
 
 void WorkerTracer::addDiagnosticChannelEvent(const tracing::InvocationSpanContext& context,
     kj::Date timestamp,
     kj::String channel,
     kj::Array<kj::byte> message) {
-  if (trace->exceededDiagnosticChannelEventLimit) {
+  if (pipelineLogLevel == PipelineLogLevel::NONE) {
     return;
   }
-  if (pipelineLogLevel == PipelineLogLevel::NONE) {
+
+  KJ_IF_SOME(writer, maybeTailStreamWriter) {
+    auto maybeTruncatedChannel = channel.first(kj::min(channel.size(), MAX_TRACE_BYTES));
+    auto maybeTruncatedMessage =
+        message.first(kj::min(message.size(), MAX_TRACE_BYTES - maybeTruncatedChannel.size()));
+    writer->report(context,
+        {tracing::DiagnosticChannelEvent(timestamp, kj::str(maybeTruncatedChannel),
+            kj::heapArray<kj::byte>(maybeTruncatedMessage))});
```

Comment:
This one should not be truncated. the value is a v8 serialized value, truncating will break the deserialization. For this one, I think it's fine to drop it on the floor if it exceeds the size limit.